### PR TITLE
KAFKA-7549: Old ProduceRequest with zstd compression does not return error to client

### PR DESCRIPTION
**Pull request by dongjinleekr**
As of current version (2.1.0), zstd-related validations are located in following spots:

1. **ProduceRequest**

    - `MemoryRecordsBuilder`: can't create `MemoryRecords` with magic < 2 (`IllegalArgumentException`)
    - `ProduceRequest.Builder`: can't create `ProduceRequest` with api version below 7 (`InvalidRecordException`)

2. **FetchRequest**

    - `KafkaApis#handleFetchRequest`: Returns `FetchResponse` w/ `Errors#UNSUPPORTED_COMPRESSION_TYPE` if ...
        1. `FetchRequest` w/ API version < 10 is delivered to a zstd-compressed topic.
        2. Down-conversion failure: `LazyDownConversionRecords#makeNext` → `RecordsUtil#downConvert` throws `UnsupportedCompressionTypeException`.

3. **Etc**

    - `AbstractLegacyRecordBatch.DeepRecordsIterator`: A boilerplate validation for legacy record batches.

**In short, there is no broker-side validation for `ProduceRequest` w/ zstd compressed records.** This PR compensates this hole.

There is a reason why this validation can't be located in other class, e.g., `LogValidator`: it can't see the API version of `ProduceRequest.` The only method that can check both of `CompressionType` and API version is `KafkaApis#handleProduceRequest`; it's why.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

File: core/src/main/scala/kafka/server/KafkaApis.scala [hunk 1]
```diff
@@ -391,20 +391,23 @@ class KafkaApis(val requestChannel: RequestChannel,
...
     val nonExistingTopicResponses = mutable.Map[TopicPartition, PartitionResponse]()
+    val invalidRequestResponses = mutable.Map[TopicPartition, PartitionResponse]()
     val authorizedRequestInfo = mutable.Map[TopicPartition, MemoryRecords]()
 
     for ((topicPartition, memoryRecords) <- produceRequest.partitionRecordsOrFail.asScala) {
       if (!authorize(request.session, Write, Resource(Topic, topicPartition.topic, LITERAL)))
         unauthorizedTopicResponses += topicPartition -> new PartitionResponse(Errors.TOPIC_AUTHORIZATION_FAILED)
       else if (!metadataCache.contains(topicPartition))
         nonExistingTopicResponses += topicPartition -> new PartitionResponse(Errors.UNKNOWN_TOPIC_OR_PARTITION)
+      else if (request.header.apiVersion() < 7 && memoryRecords.batches().asScala.exists(_.compressionType() == CompressionType.ZSTD))
```
Comment (hachikuji) [inline core/src/main/scala/kafka/server/KafkaApis.scala]: Hmm.. The `ProduceRequest` was already constructed, so wouldn't `ProduceRequest.validateRecords` already have failed? I don't see much choice but to move some of the validation logic outside of the constructor. Maybe instead we can do that validation here?

Comment (dongjinleekr) [reply core/src/main/scala/kafka/server/KafkaApis.scala]: > Hmm.. The `ProduceRequest` was already constructed, so wouldn't `ProduceRequest.validateRecords` already have failed?

Well, if the user is using the java client then `ProduceRequest` with api version < 7 compressed with zstd would be already failed. This validation logic defends the other cases - that is, for example, the 3rd party clients without the validation logic.

> I don't see much choice but to move some of the validation logic outside of the constructor. Maybe instead we can do that validation here?

I'm sorry but I can't understand what you mean exactly; The validation condition here is exactly same with that of `ProduceRequest#validateRecords` (api version < 7, compressed with zstd.) Do you mean remove the validation logic in `ProduceRequest#validateRecords` and validate here only?

Comment (hachikuji) [reply core/src/main/scala/kafka/server/KafkaApis.scala]: > I'm sorry but I can't understand what you mean exactly; The validation condition here is exactly same with that of ProduceRequest#validateRecords (api version < 7, compressed with zstd.) Do you mean remove the validation logic in ProduceRequest#validateRecords and validate here only?

Yes, sorry if I wasn't clear. As long as the validation is in the `ProduceRequest` constructor, then we will fail before the handler is invoked and we will not be able to return the right error codes. So at least for the `ProduceRequest(Struct)` constructor, I think we need to take out the call to `validateRecords` and move it here instead. Does that make sense?

Comment (dongjinleekr) [reply core/src/main/scala/kafka/server/KafkaApis.scala]: So, you mean...

1. Double validation is overkill.
2. Giving the user accurate error message has high priority.

Therefore, conclusion: **moving the validation logic from `ProducerRequest#validateRecords` into `KafkaApis#handleProduceRequest`  is better.** Since we already have validation logic in `KafkaApis#handleProduceRequest`, removing it from `ProducerRequest#validateRecords` is the right direction.

Do I understand correctly?

**Comment by dongjinleekr** [comment]
Then, is this alright?

File: core/src/main/scala/kafka/server/KafkaApis.scala (hunk 1 above)
Comment (hachikuji) [reply core/src/main/scala/kafka/server/KafkaApis.scala]: I think I am not explaining the problem very clearly. When a request is received from the network, the broker must first parse it. If there are any problems parsing the request, then the broker does not know how to return an error and simply disconnects. That is what is happening here. 

The issue is that the validation logic we have inside `ProduceRequest.validateRecords` is called during the construction of the `ProduceRequest` instance. This logic is invoked during the parsing of the request. So if any of these checks fail, then we do not return any error code at all. We just disconnect.

So to fix this problem and return a helpful error code, we have to let the `ProduceRequest` parsing succeed, which means we have to take that validation out of the constructor. Then we can do the validation here inside `KafkaApis.handleProduceRequest`.

Comment (hachikuji) [reply core/src/main/scala/kafka/server/KafkaApis.scala]: So basically, what you have in the PR now fixes the problem. However, it would be good to generalize. For example, we have additional checks in `ProduceRequest.validateRecords` which would have the same problem. Also, we were relying on `validateRecords` as a client-side sanity check. Maybe one approach is to make `ProduceRequest.validateRecords` into a static method and to invoke it separately after construction in the client and here in the request handler.

Comment (dongjinleekr) [reply core/src/main/scala/kafka/server/KafkaApis.scala]: AHA! Understood. I completely misunderstood `ProducerRequest#validateRecords` is called only in client side!

**Comment by dongjinleekr** [comment]
Here is the update. Since `ProducerRequest#validateRecords` modifies the internal state of `ProducerRequest` (i.e., `idempotent` and `transactional`), I first separated that logic into `ProducerRequest#setFlags` and then turned `ProducerRequest#validateRecords` into static method. `KafkaApis#handleProduceRequest` now validates `ProducerRequest` with `ProducerRequest#validateRecords.`

+1. Now `ProducerRequest#validateRecords` throws `UnsupportedCompressionTypeException` when it violates zstd condition.
+2. This update also includes rebasing upon the latest trunk.

**Comment by hachikuji** [comment]
@dongjinleekr Thanks, I think the fix looks good. I am wondering if it is possible to create a test case in `kafka.server.ProduceRequestTest` to verify that we receive the expected error code when zstd is used in older versions?

**Comment by dongjinleekr** [comment]
@hachikuji No problem. I will complete it by this weekend.

**Comment by dongjinleekr** [comment]
Here is the update with rebasing upon the latest trunk. In this update, I added a method to go around the validation logic for unit testing(commit #
c23db83). I know this approach is a little bit dangerous but could not find any other good alternative.

**Comment by dongjinleekr** [comment]
Here is the update with rebasing upon the latest trunk. In this update, I added a method to go around the validation logic for unit testing(commit c23db83). I know this approach is a little bit dangerous but could not find any other good alternative.

File: clients/src/main/java/org/apache/kafka/common/requests/ProduceRequest.java [hunk 2]
```diff
@@ -172,6 +173,16 @@ public Builder(short minVersion,
 
         @Override
         public ProduceRequest build(short version) {
+            return build(version, true);
+        }
+
+        public ProduceRequest build(short version, boolean validate) {
```
Comment (hachikuji) [inline clients/src/main/java/org/apache/kafka/common/requests/ProduceRequest.java]: Perhaps this can be private and we can expose a `buildUnsafe` method that sets `validate` to false. Then we will be less tempted to accidentally use the API.

**Review by hachikuji — COMMENTED** [review]
Thanks, left a minor suggestion. Otherwise looks good.

**Review by hachikuji — APPROVED** [review]
LGTM, thanks for the patch. @dongjinleekr Note that I went ahead and pushed the minor suggestion that I had.

**Comment by dongjinleekr** [comment]
@hachikuji I greatly appreciate your help. :+1:
